Spigot-Server-Patches/0167-Anti-Xray.patch

```diff
@@ -0,0 +1,1541 @@
++    public void writeBlocks(PacketDataSerializer packetDataSerializer, Chunk chunk, int y, Chunk[] nearbyChunks) {
++        packetDataSerializer.writeByte(this.e);
++        this.c.b(packetDataSerializer);
++        chunk.world.blockPacketController.writeBlocks(packetDataSerializer, chunk, y, this, this.b, nearbyChunks);
++    }
++    // Paper end
++
+     public void b(PacketDataSerializer packetdataserializer) {
+         packetdataserializer.writeByte(this.e);
+         this.c.b(packetdataserializer);
+diff --git a/src/main/java/net/minecraft/server/EntityFallingBlock.java b/src/main/java/net/minecraft/server/EntityFallingBlock.java
+index 0d8c438..a0a8dac 100644
+--- a/src/main/java/net/minecraft/server/EntityFallingBlock.java
++++ b/src/main/java/net/minecraft/server/EntityFallingBlock.java
+@@ -71,6 +71,7 @@ public class EntityFallingBlock extends Entity {
+                 blockposition = new BlockPosition(this);
+                 if (this.world.getType(blockposition).getBlock() == block && !CraftEventFactory.callEntityChangeBlockEvent(this, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.AIR, 0).isCancelled()) {
+                     this.world.setAir(blockposition);
++                    this.world.blockPacketController.updateNearbyBlocks(this.world, blockposition); // Paper - Anti-Xray
+                 } else if (!this.world.isClientSide) {
+                     this.die();
+                     return;
+@@ -117,6 +118,7 @@ public class EntityFallingBlock extends Entity {
+                                     return;
+                                 }
+                                 this.world.setTypeAndData(blockposition, this.block, 3);
++                                this.world.blockPacketController.updateNearbyBlocks(this.world, blockposition); // Paper - Anti-Xray
+                                 // CraftBukkit end
+                                 if (block instanceof BlockFalling) {
+                                     ((BlockFalling) block).a_(this.world, blockposition);
+diff --git a/src/main/java/net/minecraft/server/Explosion.java b/src/main/java/net/minecraft/server/Explosion.java
+index 0b9bc91..dc4b9ce 100644
+--- a/src/main/java/net/minecraft/server/Explosion.java
++++ b/src/main/java/net/minecraft/server/Explosion.java
+@@ -233,6 +233,7 @@ public class Explosion {
+                 blockposition = (BlockPosition) iterator.next();
+                 IBlockData iblockdata = this.world.getType(blockposition);
+                 Block block = iblockdata.getBlock();
++                this.world.blockPacketController.updateNearbyBlocks(this.world, blockposition); // Paper - Anti-Xray
+ 
+                 if (flag) {
+                     double d0 = (double) ((float) blockposition.getX() + this.world.random
```

Comment:
This needs to be an obfhelper defined on the same line as the variable c (to ensure it conflicts on updates)